### PR TITLE
support-offer-tags

### PR DESCRIPTION
**Pull request by iamanikeev**

### CODE REVIEW

File: pyoffers/models/affiliate.py [hunk 1]
```diff
@@ -1,8 +1,19 @@
 # coding: utf-8
-from .core import Model, ModelManager
+from .core import InvisibleModelManager, Model, ModelManager
 from .offer_file import OfferFileManager
 
 
+class AffiliateOffer(Model):
+    """"
```
Comment (Stranger6667) [inline pyoffers/models/affiliate.py]: Extra " should be removed

File: pyoffers/models/affiliate.py [hunk 2]
```diff
@@ -1,8 +1,19 @@
...
+    Model that defines relation between Offer and Affiliate
+    """
+    pass
```
Comment (Stranger6667) [inline pyoffers/models/affiliate.py]: `pass` is redundant

File: pyoffers/utils.py [hunk 3]
```diff
@@ -65,6 +65,23 @@ def prepend_model(self, value, model):
...
+    Contains keys as containers. Returns value if container contains given key.
+    """
+
+    def __getitem__(self, key):
+        for k, v in self.items():
+            if key in k:
+                return v
+
+    def get(self, k, d=None):
+        return self.__getitem__(k) or d
```
Comment (Stranger6667) [inline pyoffers/utils.py]: Would `return self[k] or d` work?

File: tests/models/test_tags.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
+from pyoffers.models.tags import TagRelation
+
+from pyoffers.models import Tag
+
+CASSETTE_NAME = 'tag'
+
+
+def test_create_success(tag):
+    assert isinstance(tag, Tag)
+    assert tag.name == tag.name
```
Comment (Stranger6667) [inline tests/models/test_tags.py]: Why to check attribute to equality with the same attribute?

**Review by Stranger6667 — CHANGES_REQUESTED** [review]

File: pyoffers/models/affiliate.py (hunk 2 above)
Comment (iamanikeev) [reply pyoffers/models/affiliate.py]: oops, `ValuesDict` shouldn't even be here, we have removed it in previous pull request. This is a merge mistake.

File: pyoffers/utils.py (hunk 3 above)
Comment (iamanikeev) [reply pyoffers/utils.py]: oops, `ValuesDict` shouldn't even be here, we have removed it in previous pull request. This is a merge mistake.

File: tests/models/test_tags.py (hunk 4 above)
Comment (iamanikeev) [reply tests/models/test_tags.py]: There was actual name there first, then I changed it when adjusting tests. I think this check is not necessary here at all, will remove.

File: pyoffers/models/affiliate.py (hunk 1 above)
Comment (iamanikeev) [reply pyoffers/models/affiliate.py]: removed

File: pyoffers/models/affiliate.py (hunk 2 above)
Comment (iamanikeev) [reply pyoffers/models/affiliate.py]: removed

File: pyoffers/utils.py (hunk 3 above)
Comment (iamanikeev) [reply pyoffers/utils.py]: done

File: tests/models/test_tags.py (hunk 4 above)
Comment (iamanikeev) [reply tests/models/test_tags.py]: done

File: docs/changelog.rst [hunk 5]
```diff
@@ -12,6 +12,7 @@ Changelog
 Added
 ~~~~~
 - ``generate_tracking_link`` method for ``Offer`` model. (`iamanikeev`_)
+- Support of Tag controller. (`iamanikeev`_)
```
Comment (Stranger6667) [inline docs/changelog.rst]: Don't add new changes to an existing release. It doesn't make sense. Please, add it to an appropriate section

File: pyoffers/models/__init__.py [hunk 6]
```diff
@@ -24,11 +31,17 @@
...
 MANAGER_ALIASES = {
     # Some API calls return instances of object under different name. F.e. you have to specify ``contain=['Thumbnail']``
     # in order to find Offer thumbnail (https://developers.tune.com/network/offer-findall/). But it will in fact return
     # ``OfferFile`` instance under ``Thumbnail``. For that we need aliases.
-    'Thumbnail': 'OfferFile'
+    'Thumbnail': 'OfferFile',
+    'OfferTag': 'Tag',
+    'AdvertiserTag': 'Tag',
+    'AffiliateTag': 'Tag',
+
```
Comment (Stranger6667) [inline pyoffers/models/__init__.py]: Remove this empty line, please

File: pyoffers/models/tags.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    def remove_from_affiliate(self, tag_id, affiliate_id):
+        return self._call('removeFromAffiliate', tag_id=tag_id, affiliate_id=affiliate_id)
+
+    def remove_from_advertiser(self, tag_id, advertiser_id):
+        return self._call('removeFromAdvertiser', tag_id=tag_id, advertiser_id=advertiser_id)
+
+    def remove_from_advertiser_by_relational_id(self, id):
+        return self._call('removeFromAdvertiserByRelationalId', id=id)
+
+    # todo: uncomment when HO fixes bugs in this method
```
Comment (Stranger6667) [inline pyoffers/models/tags.py]: Please, don't add commented code. Either add it or not.
I'd suggest to add it

File: docs/changelog.rst (hunk 5 above)
Comment (iamanikeev) [reply docs/changelog.rst]: Sorry, this is some kind of optical illusion in my brain, I always miss the latest release section and think of latest `Added` section a part of `Unreleased`. Fixed.

File: pyoffers/models/__init__.py (hunk 6 above)
Comment (iamanikeev) [reply pyoffers/models/__init__.py]: done

File: pyoffers/models/tags.py (hunk 7 above)
Comment (iamanikeev) [reply pyoffers/models/tags.py]: Added. Marked corresponding tests as skipped.

**Pull request merged**